### PR TITLE
Plane/Ray add copy/clone methods + prefer ownership over references

### PR DESCRIPTION
**Pull request by kungfooman**
We are preaching that devs should use `Vec3.FORWARD` etc. but then these two classes don't even support it.

This also adds `copy` and `clone` methods that I always expect (but then I run into script errors).

I confirm I have read the [contributing guidelines](https://github.com/playcanvas/engine/blob/master/.github/CONTRIBUTING.md) and signed the [Contributor License Agreement](https://docs.google.com/a/playcanvas.com/forms/d/1Ih69zQfJG-QDLIEpHr6CsaAs6fPORNOVnMv5nuo0cjk/viewform).

### CODE REVIEW

**Comment by yaustar** [comment]
I wonder if references were done for optimisation purposes 🤔 We also have to be careful here as this pattern is done across the bounding shapes like AABB, orientated box and sphere.

**Comment by kungfooman** [comment]
> I wonder if references were done for optimisation purposes

I rewrote it a little bit for comparing the ES5 transpilation.

How it looks transpiled to ES5 right now without PR:

```js
function Plane(point, normal) {
    if (point === void 0) {
      point = new Vec3();
    }
    if (normal === void 0) {
      normal = new Vec3(0, 0, 1);
    }
    this.point = point;
    this.normal = normal;
}
```

The way it looks transpiled to ES5 with PR:

```js
function Plane(point, normal) {
    this.point = Vec3.ZERO.clone();
    this.normal = Vec3.BACK.clone();
    if (point) {
        this.point.copy(point);
    }
    if (normal) {
        this.normal.copy(normal);
    }
}
```

Then I wondered how often the references are actually used (tldr: never):

image


For plane its 1x used without references:

image

I don't think this is a CPU performance drain.

Basically I would just like these Math classes to be simpler for two cases:

 - It's easy to forget to clone the input arguments and then the Plane/Ray instances are just referencing instances from completely different classes (e.g. `GraphNode#localPosition`).

- Allow the use of `Vec3.FORWARD` etc. to be in line with the rest of the engine "style decisions"

Thank you for bringing in the other classes like `BoundingBox`, imo they should rather "own" their data aswell, but it's nice to simply brainstorm this first for pros/cons.

**Comment by mvaligursky** [comment]
I fully agree a copy is lot better, and it should not cause any problems for the Plane and the Ray, so I'm happy to do this.
I've tried doing this for all shapes in the past, but hit some problems where some engine code depends on references (and those are hard to find), so this never got merged. Since then I've removed some of those dependencies on references, but there might still be some.

https://github.com/playcanvas/engine/pull/2088

**Comment by kungfooman** [comment]
Wow, that PR is back in ES5 :sweat_smile: 

> Since then I've removed some of those dependencies on references, but there might still be some.

Yea, there are cases a reference makes sense and it feels to me that should rather be quite a "special case". There could be a `static Plane#fromReferences` method for example? That would be as explicit as it gets to mark such cases (and easy to search over hundreds of files).

**Comment by mvaligursky** [comment]
Personally, I'd be a lot more happy to not use references for the shape members at all, the copy is not that costly, and those do not tend to get constructed on the fly anyways.

File: src/core/shape/plane.js [hunk 1]
```diff
@@ -8,17 +8,46 @@ const tmpVecA = new Vec3();
  * @ignore
  */
 class Plane {
+    /**
+     * The starting point of the plane.
+     *
+     * @type {Vec3}
+     */
+    point = Vec3.ZERO.clone();
```
Comment (mvaligursky) [inline src/core/shape/plane.js:16]:
```suggestion
    point = new Vec3();
```

File: src/core/shape/ray.js [hunk 2]
```diff
@@ -1,39 +1,43 @@
...
 /**
  * An infinite ray.
  */
 class Ray {
+    /**
+     * The starting point of the ray.
+     *
+     * @type {Vec3}
+     */
+    origin = Vec3.ZERO.clone();
```
Comment (mvaligursky) [inline src/core/shape/ray.js:12]: we should add '@ignore' tag on the origin and direction members, to hide a public API allowing user to swap those to other instances of Vec3

Comment (mvaligursky) [inline src/core/shape/ray.js:12]:
```suggestion
    origin = new Vec3();
```

Comment (kungfooman) [reply src/core/shape/ray.js]: Thank you for your review @mvaligursky, I like your simplification! In this case, I am not sure, do you mean `@readonly`?

image

(it allows read-access while not just dropping it from docs)

Comment (mvaligursky) [reply src/core/shape/ray.js]: ah yeah, readonly is even better!

Comment (kungfooman) [reply src/core/shape/ray.js]: Thank you a lot for the annotation hint! It feels much better now that Intellisense points it out.

**Review by mvaligursky — APPROVED** [review]
Looks great, thanks for the contri!

**Comment by willeastcott** [comment]
This PR generates the following error:

image

This is solved in other classes as follows:

https://github.com/playcanvas/engine/blob/e477bd13b42a9f548153fe33df84430d891cf3a9/src/core/math/vec3.js#L119-L121

**Comment by kungfooman** [comment]
@willeastcott It's not really solved, just dragging TypeScript's incompetence of understanding basic OOP concepts over more lines:

image

**Pull request merged**